### PR TITLE
Subscribers: Redirect user if insufficient privileges

### PR DESCRIPTION
**Pull request by danielbachhuber**
Related https://github.com/Automattic/dotcom-forge/issues/4708

## Proposed Changes

Redirects the user to My Home if they don't have sufficient privileges to visit Subscribers.

## Testing Instructions

1. Create a new Business site.
2. With the site owner, visit `/subscribers/<site>` and verify the page loads as expected.
3. Add a second user to the site as an author.
4. With the second user, visit `/subscribers/<site>` and verify you're redirected to My Home.

### CODE REVIEW

**Review by jeroenpf — APPROVED** [review]
Code looks good and works as expected.

Followed the testing instructions and a privileged user is not redirected while an unprivileged user is.

The `redirectIfInsufficientPrivileges` middleware was also added here: https://github.com/Automattic/wp-calypso/pull/85913

Do we perhaps want to make this into something reusable so that we don't have to copy/rewrite this piece of code every time we want to add middleware that checks for a specific set of privileges? 

Could be something like `redirectifInsufficientPrivileges( [ 'manage_options'], optional_path )`

This is nothing critical or blocking so I've approved this PR.

**Comment by wojtekn** [comment]
> Do we perhaps want to make this into something reusable so that we don't have to copy/rewrite this piece of code every time we want to add middleware that checks for a specific set of privileges?

This sounds like a good idea. We keep common middlewares in `client/controller/index.web.js`.

File: client/my-sites/subscribers/controller.js [hunk 1]
```diff
@@ -23,6 +25,21 @@ const queryStringChanged = ( key ) => ( value ) => {
 	return page.show( addQueryArgs( path, { [ key ]: value } ) );
 };
 
+export async function redirectIfInsufficientPrivileges( context, next ) {
+	const { store } = context;
+	const state = store.getState();
+	const site = getSelectedSite( state );
+	const canManageOptions = canCurrentUser( state, site.ID, 'manage_options' );
```
Comment (simison) [inline client/my-sites/subscribers/controller.js]: There's a bunch of `*_users` capabilities, should we use `list_users` here instead?

https://wordpress.org/documentation/article/roles-and-capabilities/

Comment (simison) [reply client/my-sites/subscribers/controller.js]: Not a blocker for merging, to be clear. `manage_options` is fine for me.

Comment (danielbachhuber) [reply client/my-sites/subscribers/controller.js]: > There's a bunch of `*_users` capabilities, should we use `list_users` here instead?

You're right. `list_users` is what we use elsewhere: https://github.com/Automattic/jetpack/blob/bbd17c20e9f2659701f639c84124e3cfa1a33de2/projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php#L131

Updated in ebf7e623a57ec556b668f4886e171189bf88cdfb

**Comment by danielbachhuber** [comment]
> Do we perhaps want to make this into something reusable so that we don't have to copy/rewrite this piece of code every time we want to add middleware that checks for a specific set of privileges?

@jeroenpf Good suggestion. Refactored this in ebf7e623a57ec556b668f4886e171189bf88cdfb

File: client/controller/index.node.js [hunk 2]
```diff
@@ -82,6 +82,8 @@ export const redirectLoggedOut = () => {};
 export const redirectLoggedOutToSignup = () => {};
 export const redirectWithoutLocaleParamIfLoggedIn = () => {};
 // eslint-disable-next-line no-unused-vars
+export const redirectIfCurrentUserCannot = ( capability ) => {};
```
Comment (danielbachhuber) [inline client/controller/index.node.js]: Is this necessary/appropriate?

Comment (simison) [reply client/controller/index.node.js]: @tyxla ? :-)

Comment (danielbachhuber) [reply client/controller/index.node.js]: @Automattic/team-calypso-frameworks

Comment (tyxla) [reply client/controller/index.node.js]: Looks good. We could likely ditch the unused argument (and the ESLint ignore) though, I don't think there's a good reason to keep it.

**Review by wojtekn — APPROVED** [review]

File: client/my-sites/subscribers/index.js [hunk 3]
```diff
@@ -1,15 +1,36 @@
...
 
 export default function () {
-	page( '/subscribers', siteSelection, sites, navigation, makeLayout, clientRender );
-	page( '/subscribers/:domain', siteSelection, navigation, subscribers, makeLayout, clientRender );
+	page(
+		'/subscribers',
+		siteSelection,
+		sites,
+		navigation,
+		redirectIfCurrentUserCannot( 'list_users' ),
```
Comment (Aurorum) [inline client/my-sites/subscribers/index.js:16]: This appears to cause a crash when there's no site ID - eg. https://wordpress.com/subscribers

I've tried to fix this in #86556, if you'd be okay with taking a look at it. :)

**Pull request merged**